### PR TITLE
Feat/ram reservation

### PR DESCRIPTION
**Pull request by DK101010**
### Description

WIth this PR user can manuelly modify the ram reservation for a specific vm in vmware context. User set it via vm settings as a float value between 0 and 1

### Types of changes

- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] Enhancement (improves an existing feature and functionality)
- [ ] Cleanup (Code refactoring and cleanup, that may add test cases)

#### Feature/Enhancement Scale

- [ ] Major
- [x] Minor

#### Bug Severity

- [ ] BLOCKER
- [ ] Critical
- [ ] Major
- [ ] Minor
- [ ] Trivial

### Screenshots (if appropriate):
image

### How Has This Been Tested?
tested in own vmware environment

### CODE REVIEW

File: vmware-base/src/main/java/com/cloud/hypervisor/vmware/util/VmwareHelper.java [hunk 1]
```diff
@@ -549,12 +549,6 @@ public static void setBasicVmConfig(VirtualMachineConfigSpec vmConfig, int cpuCo
 
         cpuInfo.setReservation((long)cpuReservedMhz);
         vmConfig.setCpuAllocation(cpuInfo);
-        if (cpuSpeedMHz != cpuReservedMhz) {
-            vmConfig.setCpuHotAddEnabled(true);
-        }
-        if (memoryMB != memoryReserveMB) {
-            vmConfig.setMemoryHotAddEnabled(true);
-        }
```
Comment (DaanHoogland) [inline vmware-base/src/main/java/com/cloud/hypervisor/vmware/util/VmwareHelper.java:557]: thsi removal does not seem relevant to your feature description. can you explain @DK101010 ?

Comment (DK101010) [reply vmware-base/src/main/java/com/cloud/hypervisor/vmware/util/VmwareHelper.java:557]: @DaanHoogland sorry for my late answer. In my opinion I think yes and no. 
Hot add memory and cpu will be enabled when reserved memory and cpu is in use. But this is not correct. Both features are independent, therefore I thought it is good to remove it in this PR. But if you want I can create a separate PR for this commit.

Comment (DaanHoogland) [reply vmware-base/src/main/java/com/cloud/hypervisor/vmware/util/VmwareHelper.java:557]: not sure if that is needed and/or suffices; How about when people want to hot-add and don't care about reservation?

Comment (DK101010) [reply vmware-base/src/main/java/com/cloud/hypervisor/vmware/util/VmwareHelper.java:557]: @DaanHoogland hot add and reservation are two different features and they are independent of each other. I talked with my team and test it in my environment to get sure and in my env it works. In my point of view this code lines makes no sense.

Comment (DaanHoogland) [reply vmware-base/src/main/java/com/cloud/hypervisor/vmware/util/VmwareHelper.java:557]: ok @DK101010 , if that's verified I'm fine with the code.

**Comment by DaanHoogland** [comment]
@blueorangutan package

**Comment by DaanHoogland** [comment]
@blueorangutan package

**Comment by DaanHoogland** [comment]
@blueorangutan test centos7 vmware-67u3

File: plugins/hypervisors/vmware/src/main/java/com/cloud/hypervisor/vmware/resource/VmwareResource.java [hunk 2]
```diff
@@ -2905,7 +2905,7 @@ private void tearDownVm(VirtualMachineMO vmMo) throws Exception {
 
     int getReservedMemoryMb(VirtualMachineTO vmSpec) {
         if (vmSpec.getDetails().get(VMwareGuru.VmwareReserveMemory.key()).equalsIgnoreCase("true")) {
-            if(!vmSpec.getDetails().get(VmDetailConstants.RAM_RESERVATION).equals("0.0")){
+            if(vmSpec.getDetails().get(VmDetailConstants.RAM_RESERVATION) != null && !vmSpec.getDetails().get(VmDetailConstants.RAM_RESERVATION).equals("0.0")){
```
Comment (DaanHoogland) [inline plugins/hypervisors/vmware/src/main/java/com/cloud/hypervisor/vmware/resource/VmwareResource.java:2908]: how about 
```suggestion
            if(!"0.0".equals(vmSpec.getDetails().get(VmDetailConstants.RAM_RESERVATION))){
```

**Comment by DaanHoogland** [comment]
@GabrielBrascher @sureshanaparti @nvazquez (not sure who else to bother), can you guys look at this?
tnx

File: vmware-base/src/main/java/com/cloud/hypervisor/vmware/util/VmwareHelper.java (hunk 1 above)
Comment (GabrielBrascher) [reply vmware-base/src/main/java/com/cloud/hypervisor/vmware/util/VmwareHelper.java:557]: I have not much experience with the codebase for VMware to comment on this piece of code. Good that it has been verified. I will rely on the contributor tests.

File: plugins/hypervisors/vmware/src/main/java/com/cloud/hypervisor/guru/VmwareVmImplementer.java [hunk 3]
```diff
@@ -139,6 +140,9 @@ VirtualMachineTO implement(VirtualMachineProfile vm, VirtualMachineTO to, long c
                     details.put(VmDetailConstants.NIC_ADAPTER, VirtualEthernetCardType.E1000.toString());
                 }
             }
+            if(StringUtils.isEmpty(details.get(VmDetailConstants.RAM_RESERVATION))){
+                details.put(VmDetailConstants.RAM_RESERVATION, "0.0");
```
Comment (GabrielBrascher) [inline plugins/hypervisors/vmware/src/main/java/com/cloud/hypervisor/guru/VmwareVmImplementer.java:144]: There are two occurrences of the "0.0" String.

What do you think of replacing these `"0.0"` with `NumberUtils.DOUBLE_ZERO`? The NumberUtils will return a `Double` object of `0.0d` and, therefore, it has a `toString()` that returns the String `"0.0"`.

Other option would be to create a constant with a self-describing name.

**Review by GabrielBrascher — COMMENTED** [review]
Added a small suggestion.

File: plugins/hypervisors/vmware/src/main/java/com/cloud/hypervisor/guru/VmwareVmImplementer.java (hunk 3 above)
Comment (DK101010) [reply plugins/hypervisors/vmware/src/main/java/com/cloud/hypervisor/guru/VmwareVmImplementer.java:144]: Hi @GabrielBrascher, sorry for my late answer, I must have missed the notification.  I think there are no benefits regards reusable of this code, but I agree it's better to use parameter instead of strings. I'm prefer your first suggestion and will adapt  it.

**Review by DaanHoogland — APPROVED** [review]
clgtm, test failure is does not seem related

File: vmware-base/src/main/java/com/cloud/hypervisor/vmware/util/VmwareHelper.java (hunk 1 above)
Comment (rhtyd) [reply vmware-base/src/main/java/com/cloud/hypervisor/vmware/util/VmwareHelper.java:557]: I've my doubts on these changes, pinged vmware experts @harikrishna-patnala @nvazquez @sureshanaparti

**Comment by nvazquez** [comment]
@blueorangutan package

**Comment by nvazquez** [comment]
@blueorangutan test

**Comment by nvazquez** [comment]
@davidjumani can you please test this PR? I think this will only apply for non-deploy-as-is VMs, but not for deploy-as-is

**Review by davidjumani — APPROVED** [review]
Screenshot from 2021-07-21 16-24-51

File: plugins/hypervisors/vmware/src/main/java/com/cloud/hypervisor/guru/VmwareVmImplementer.java [hunk 4]
```diff
@@ -139,6 +141,9 @@ VirtualMachineTO implement(VirtualMachineProfile vm, VirtualMachineTO to, long c
                     details.put(VmDetailConstants.NIC_ADAPTER, VirtualEthernetCardType.E1000.toString());
                 }
             }
+            if(StringUtils.isEmpty(details.get(VmDetailConstants.RAM_RESERVATION))){
```
Comment (davidjumani) [inline plugins/hypervisors/vmware/src/main/java/com/cloud/hypervisor/guru/VmwareVmImplementer.java:144]: Could cause an issue if the user does not want to reserve memory for a VM, better to leave it unset and check whether the key exists in getReservedMemoryMb

Comment (DK101010) [reply plugins/hypervisors/vmware/src/main/java/com/cloud/hypervisor/guru/VmwareVmImplementer.java:144]: Hi @davidjumani , thanks for testing. It strange that it not work this line in getReservedMemory should prevent this 
```java
if(!NumberUtils.DOUBLE_ZERO.toString().equals(vmSpec.getDetails().get(VmDetailConstants.RAM_RESERVATION))){
```
But the Method returns every time the minimum ram. 
```java
return (int) (vmSpec.getMinRam() / ResourceType.bytesToMiB);
```
Could it be in your test case that the reserved memory = minimum ram is.

Comment (davidjumani) [reply plugins/hypervisors/vmware/src/main/java/com/cloud/hypervisor/guru/VmwareVmImplementer.java:144]: Nope, passing 0 ensures that no memory is reserved, but 0.0 does not. When 0.0 is passed, all the memory is reserved

Comment (DK101010) [reply plugins/hypervisors/vmware/src/main/java/com/cloud/hypervisor/guru/VmwareVmImplementer.java:144]: Strange, I'll take care of it. Thanks for the infos.

**Comment by davidjumani** [comment]
@blueorangutan package

File: plugins/hypervisors/vmware/src/main/java/com/cloud/hypervisor/guru/VmwareVmImplementer.java [hunk 5]
```diff
@@ -57,6 +57,8 @@
 import java.util.HashMap;
 import java.util.List;
 import java.util.Map;
+import org.apache.commons.lang.StringUtils;
+import org.apache.commons.lang3.math.NumberUtils;
```
Comment (davidjumani) [inline plugins/hypervisors/vmware/src/main/java/com/cloud/hypervisor/guru/VmwareVmImplementer.java:61]:
```suggestion
```

**Comment by davidjumani** [comment]
@blueorangutan package

**Comment by nvazquez** [comment]
@blueorangutan test centos7 vmware-67u3

File: plugins/hypervisors/vmware/src/main/java/com/cloud/hypervisor/vmware/resource/VmwareResource.java [hunk 6]
```diff
@@ -2906,7 +2905,10 @@ private void tearDownVm(VirtualMachineMO vmMo) throws Exception {
 
     int getReservedMemoryMb(VirtualMachineTO vmSpec) {
         if (vmSpec.getDetails().get(VMwareGuru.VmwareReserveMemory.key()).equalsIgnoreCase("true")) {
-            return (int) (vmSpec.getMinRam() / ResourceType.bytesToMiB);
+            if(vmSpec.getDetails().get(VmDetailConstants.RAM_RESERVATION) != null){
+                float reservedMemory = (vmSpec.getMaxRam() * Float.parseFloat(vmSpec.getDetails().get(VmDetailConstants.RAM_RESERVATION)));
+                return (int) (reservedMemory / ResourceType.bytesToMiB);
+            }
```
Comment (davidjumani) [inline plugins/hypervisors/vmware/src/main/java/com/cloud/hypervisor/vmware/resource/VmwareResource.java:2911]:
```suggestion
            }
            return (int) (vmSpec.getMinRam() / ResourceType.bytesToMiB);
```

By default reserve all memory

**Comment by nvazquez** [comment]
@blueorangutan package

**Comment by nvazquez** [comment]
@blueorangutan test

File: vmware-base/src/main/java/com/cloud/hypervisor/vmware/util/VmwareHelper.java (hunk 1 above)
Comment (nvazquez) [reply vmware-base/src/main/java/com/cloud/hypervisor/vmware/util/VmwareHelper.java:557]: @DK101010 I think removing these lines will cause a regression on cpu and memory overcommit, have you tested those cases?

File: vmware-base/src/main/java/com/cloud/hypervisor/vmware/util/VmwareHelper.java [hunk 7]
```diff
@@ -549,12 +549,6 @@ public static void setBasicVmConfig(VirtualMachineConfigSpec vmConfig, int cpuCo
 
         cpuInfo.setReservation((long)cpuReservedMhz);
         vmConfig.setCpuAllocation(cpuInfo);
-        if (cpuSpeedMHz != cpuReservedMhz) {
```
Comment (nvazquez) [inline vmware-base/src/main/java/com/cloud/hypervisor/vmware/util/VmwareHelper.java:552]: @DK101010 I think removing these lines will cause a regression on cpu and memory overcommit, have you tested those cases?

**Review by nvazquez — COMMENTED** [review]
Only one concern, besides it it looks good

File: vmware-base/src/main/java/com/cloud/hypervisor/vmware/util/VmwareHelper.java (hunk 1 above)
Comment (DK101010) [reply vmware-base/src/main/java/com/cloud/hypervisor/vmware/util/VmwareHelper.java:557]: Hi @nvazquez not sure what do you mean with regression.
Reservation and hot add are two different things. I can use reservation in it doesn't matter if hot add enabled/disbaled and it makes for me not sense to enable this when I'm reserve memory .
Also you can't more reserve as you have for the current vm. 

I may be missing something, but currently I don't see a problem here.

File: vmware-base/src/main/java/com/cloud/hypervisor/vmware/util/VmwareHelper.java (hunk 7 above)
Comment (DK101010) [reply vmware-base/src/main/java/com/cloud/hypervisor/vmware/util/VmwareHelper.java:552]: Hi @nvazquez not sure what do you mean with regression.
Reservation and hot add are two different things. I can use reservation in it doesn't matter if hot add enabled/disbaled and it makes for me not sense to enable this when I'm reserve memory .
Also you can't more reserve as you have for the current vm.

I may be missing something, but currently I don't see a problem here.

Comment (nvazquez) [reply vmware-base/src/main/java/com/cloud/hypervisor/vmware/util/VmwareHelper.java:552]: @DK101010 I have examined the `setBasicVmConfig(..)` method is used on:
- `VmwareResource.execute(StartCommand cmd)` -> the values set on these removed lines are not relevant, are overridden later (https://github.com/apache/cloudstack/blob/main/plugins/hypervisors/vmware/src/main/java/com/cloud/hypervisor/vmware/resource/VmwareResource.java#L1995-L2005)
- `HypervisorHostHelper.createBlankVm(..)` -> possible regression at this point: blank/worker VMs are always created with `cpuHotAdd = false` and `memHotAdd = false`. I'm not sure how/if this would affect deployments, any thoughts? 

cc. @sureshanaparti @rhtyd @DaanHoogland

**Comment by nvazquez** [comment]
@blueorangutan package

**Comment by nvazquez** [comment]
@blueorangutan test centos7 vmware-67u3

File: vmware-base/src/main/java/com/cloud/hypervisor/vmware/util/VmwareHelper.java (hunk 7 above)
Comment (sureshanaparti) [reply vmware-base/src/main/java/com/cloud/hypervisor/vmware/util/VmwareHelper.java:552]: @nvazquez Hot-adding cpu & memory are set / overridden based on guest OS support. Also, the reservation and hot add are different things. So, I think, the hot add config here seems not relevant now.

**Comment by nvazquez** [comment]
@blueorangutan test centos7 vmware-67u3

File: vmware-base/src/main/java/com/cloud/hypervisor/vmware/util/VmwareHelper.java (hunk 7 above)
Comment (nvazquez) [reply vmware-base/src/main/java/com/cloud/hypervisor/vmware/util/VmwareHelper.java:552]: Thanks @sureshanaparti

**Comment by nvazquez** [comment]
@blueorangutan test centos7 vmware-67u3

**Comment by nvazquez** [comment]
@blueorangutan package

**Comment by nvazquez** [comment]
@blueorangutan test centos7 vmware-67u3

**Review by nvazquez — APPROVED** [review]
